# preprocessing text data for llms 

In [6]:
from datasets import load_dataset


In [7]:
# Save the dataset to disk
dataset.save_to_disk(r"C:\LLM from Scratch\datasets\tiny_stories_dataset")


Saving the dataset (1/1 shards): 100%|██████████| 21990/21990 [00:00<00:00, 1250088.03 examples/s]


In [10]:
dataset

DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 2119719
    })
    validation: Dataset({
        features: ['text'],
        num_rows: 21990
    })
})

### createing a small dataset and trying to split it using `re`  

In [15]:
import re
text = "This is a` sample text with ,some punctuation! Let's clean it up."
result = re.split(r'[^\w\s`]', text)

print(result)

['This is a` sample text with ', 'some punctuation', ' Let', 's clean it up', '']


In [17]:
result = re.split(r'([,.]|\s)', text)
result  = [word for word in result if word.strip() != '']
print(result)


['This', 'is', 'a`', 'sample', 'text', 'with', ',', 'some', 'punctuation!', "Let's", 'clean', 'it', 'up', '.']


lets apply this kind of tokenizer to dataset

#### our dataset is an list of strings so lets join our all the list into single string and convert it into tokens 

In [28]:
dataset_text = " ".join(dataset["train"]['text'][:20000])  # Join the first 20,000 entries of the 'text' column into a single string

In [29]:
preprocessed_text = re.split(r'([,.:;?_!"()\']|--|\s)', dataset_text)
preprocessed_text = [word for word in preprocessed_text if word.strip() != '']
print(preprocessed_text[:100])

['One', 'day', ',', 'a', 'little', 'girl', 'named', 'Lily', 'found', 'a', 'needle', 'in', 'her', 'room', '.', 'She', 'knew', 'it', 'was', 'difficult', 'to', 'play', 'with', 'it', 'because', 'it', 'was', 'sharp', '.', 'Lily', 'wanted', 'to', 'share', 'the', 'needle', 'with', 'her', 'mom', ',', 'so', 'she', 'could', 'sew', 'a', 'button', 'on', 'her', 'shirt', '.', 'Lily', 'went', 'to', 'her', 'mom', 'and', 'said', ',', '"', 'Mom', ',', 'I', 'found', 'this', 'needle', '.', 'Can', 'you', 'share', 'it', 'with', 'me', 'and', 'sew', 'my', 'shirt', '?', '"', 'Her', 'mom', 'smiled', 'and', 'said', ',', '"', 'Yes', ',', 'Lily', ',', 'we', 'can', 'share', 'the', 'needle', 'and', 'fix', 'your', 'shirt', '.', '"', 'Together']


In [30]:
print(f"Total number of tokens: {len(preprocessed_text)}")

Total number of tokens: 4236944


## converting token into token id

Next, let’s convert these tokens from a Python string to an integer representation to
produce the token IDs. This conversion is an intermediate step before converting the
token IDs into embedding vectors.


let’s create a list of all unique tokens and sort them
alphabetically to determine the vocabulary size:

In [31]:
all_words = sorted(set(preprocessed_text))
vocab_size = len(all_words)
print(f"Vocabulary size: {vocab_size}")

Vocabulary size: 13395


we have total `13395` unique words among `4236944` tokens 


## creating the Vocabulary

lets print first 100 vocabulary 

In [36]:
vocab = {word: idx for word, idx in enumerate(all_words)}
for i, item in enumerate(list(vocab.items())[:10]):
    print(item)
    if i >= 101:
        break

(0, '!')
(1, '"')
(2, '$10')
(3, '$500')
(4, '&')
(5, "'")
(6, '(')
(7, ')')
(8, '*beep*')
(9, '*poof*')
